### PR TITLE
Switch IPMI watchdog timer module to use policy for configuration

### PR DESCRIPTION
**Pull request by cfernald**
## Description

Switch the IPMI watchdog timer modules to use a policy blob instead of dynamic PCDs. This creates a stronger guarantee of finalized data availability. 

- [x] Impacts functionality?
- [ ] Impacts security?
- [x] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

Tested locally on QEMU. 

## Integration Instructions

Remove IPMI weatchdog PCDs and introduce platform code to produce watchdog timer policy.

### CODE REVIEW

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+typedef struct _IPMI_WATCHDOG_POLICY {
+  BOOLEAN    Frb2Enabled;
+  UINT8      Frb2TimeoutAction;
+  UINT16     Frb2TimeoutSeconds;
+  BOOLEAN    OsWatchdogEnabled;
+  UINT8      OsTimeoutAction;
+  UINT16     OsTimeoutSeconds;
+} IPMI_WATCHDOG_POLICY;
+
```
Comment (os-d) [inline IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:22]: I've been seeing policy headers go under `Include/Guid/*Policy.h` and include the policy guid in them. I don't have a big preference here, but just if we want to align on something or not.

**Review by os-d — APPROVED** [review]

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h (hunk 1 above)
Comment (kuqin12) [reply IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:22]: We probably should still include `gIpmiWatchdogPolicyGuid` here.

**Review by kuqin12 — APPROVED** [review]

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h [hunk 2]
```diff
@@ -0,0 +1,23 @@
+/** @file
+  Definitions for the IPMI watchdog timers policy.
+
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _IPMI_WATCHDOG_POLICY_H_
```
Comment (apop5) [inline IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:9]:
```suggestion
#ifndef IPMI_WATCHDOG_POLICY_H_

```

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/** @file
+  Definitions for the IPMI watchdog timers policy.
+
+  Copyright (c) Microsoft Corporation
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _IPMI_WATCHDOG_POLICY_H_
+#define _IPMI_WATCHDOG_POLICY_H_
```
Comment (apop5) [inline IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:10]:
```suggestion
#define IPMI_WATCHDOG_POLICY_H_

```

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+
+**/
+
+#ifndef _IPMI_WATCHDOG_POLICY_H_
+#define _IPMI_WATCHDOG_POLICY_H_
+
+#include <uefi.h>
+
+typedef struct _IPMI_WATCHDOG_POLICY {
```
Comment (apop5) [inline IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:14]: probably want to pack the structure.

File: IpmiFeaturePkg/IpmiWatchdog/Dxe/WatchdogDxe.c [hunk 5]
```diff
@@ -99,6 +102,14 @@ IpmiWatchdogDxeEntryPoint (
 {
   EFI_STATUS                        Status;
   IPMI_GET_WATCHDOG_TIMER_RESPONSE  WatchdogTimer;
+  UINT16                            PolicySize;
+
+  PolicySize = sizeof (IPMI_WATCHDOG_POLICY);
+  Status     = GetPolicy (&gIpmiWatchdogPolicyGuid, NULL, &mWatchdogPolicy, &PolicySize);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a: Failed to get watchdog policy! %r\n", __FUNCTION__, Status));
```
Comment (apop5) [inline IpmiFeaturePkg/IpmiWatchdog/Dxe/WatchdogDxe.c:110]: maybe add an assert here as well?

**Review by apop5 — APPROVED** [review]

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  BOOLEAN    Frb2Enabled;
```
Comment (apop5) [inline IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:15]: With the PCDs being removed, we also lose the comments around what each field means. Could we copy the comments from the PCDs after each policy entry?

**Comment by MarcChen46** [comment]
Please also update the [Ipmi_Watchdog.md](https://github.com/microsoft/mu_feature_ipmi/blob/main/IpmiFeaturePkg/Docs/Ipmi_Watchdog.md), and I can see one requirement is that platform code needs to implement policy creator that set the policy with finalize attribute before the driver get dispatched, for better consume this feature without extra implementation, can we also implement a default policy creator that set the finalize attribute? then for consumers that want to consume the feature directly, they do not to do extra implementation and can just use the default policy value directly.

File: IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h (hunk 1 above)
Comment (cfernald) [reply IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:22]: As a matter of contention, we should probably decide on a standard. I don't particularly care if this live ins /Guid/, but figured /Policy/ might add some clarity. Thoughts?

Comment (cfernald) [reply IpmiFeaturePkg/Include/Policy/IpmiWatchdogPolicy.h:22]: As a matter of contention, we should probably decide on a standard. I don't particularly care if this live ins /Guid/, but figured /Policy/ might add some clarity. Thoughts?

**Comment by cfernald** [comment]
> Please also update the [Ipmi_Watchdog.md](https://github.com/microsoft/mu_feature_ipmi/blob/main/IpmiFeaturePkg/Docs/Ipmi_Watchdog.md), and I can see one requirement is that platform code needs to implement policy creator that set the policy with finalize attribute before the driver get dispatched, for better consume this feature without extra implementation, can we also implement a default policy creator that set the finalize attribute? then for consumers that want to consume the feature directly, they do not to do extra implementation and can just use the default policy value directly.

I will update the readme. @os-d, @kuqin12, @apop5, and I actually had a lengthy discussion about concept of default providers. I personally want to avoid the bloat of per-policy default finalized providers as I think it is antithetical to the purpose of swapping to policy and just leads to code/binary bloat. Our current decision was to hold off on the default provider and if one seems necessary then we should evaluate a framework for default policies in the policy service rather then pushing boiler plate code to the service consumers.

**Comment by MarcChen46** [comment]
> > Please also update the [Ipmi_Watchdog.md](https://github.com/microsoft/mu_feature_ipmi/blob/main/IpmiFeaturePkg/Docs/Ipmi_Watchdog.md), and I can see one requirement is that platform code needs to implement policy creator that set the policy with finalize attribute before the driver get dispatched, for better consume this feature without extra implementation, can we also implement a default policy creator that set the finalize attribute? then for consumers that want to consume the feature directly, they do not to do extra implementation and can just use the default policy value directly.
> 
> I will update the readme. @os-d, @kuqin12, @apop5, and I actually had a lengthy discussion about concept of default providers. I personally want to avoid the bloat of per-policy default finalized providers as I think it is antithetical to the purpose of swapping to policy and just leads to code/binary bloat. Our current decision was to hold off on the default provider and if one seems necessary then we should evaluate a framework for default policies in the policy service rather then pushing boiler plate code to the service consumers. @MarcChen46, let me know if you still have concerns.

Hi @cfernald, thanks for your reply, I think my concern is still from consumer's point of view, once the policy design keeps integrating to more and more features, and if today we have 30 features that is controlled by policy of each feature, then platform code needs to implement Policy Creators for each of the feature, the effort is a little bit big to developers that want to consume MU features, and as a consumer, we may just want to consume the feature with its default behavior, in that case the original PCDs design is even easier for consumer to consume the feature, that can reduce a lot of consumer's integrating effort when consume MU features.
If we can evaluate a framework for default policies in the policy service, I think that will be the best solution, and if both Policy and PCD control design will be co-existed in MU, maybe we should also identify a best practice to suggest when to use PCD and when to use Policy design.

**Comment by cfernald** [comment]
> > > Please also update the [Ipmi_Watchdog.md](https://github.com/microsoft/mu_feature_ipmi/blob/main/IpmiFeaturePkg/Docs/Ipmi_Watchdog.md), and I can see one requirement is that platform code needs to implement policy creator that set the policy with finalize attribute before the driver get dispatched, for better consume this feature without extra implementation, can we also implement a default policy creator that set the finalize attribute? then for consumers that want to consume the feature directly, they do not to do extra implementation and can just use the default policy value directly.
> > 
> > 
> > I will update the readme. @os-d, @kuqin12, @apop5, and I actually had a lengthy discussion about concept of default providers. I personally want to avoid the bloat of per-policy default finalized providers as I think it is antithetical to the purpose of swapping to policy and just leads to code/binary bloat. Our current decision was to hold off on the default provider and if one seems necessary then we should evaluate a framework for default policies in the policy service rather then pushing boiler plate code to the service consumers. @MarcChen46, let me know if you still have concerns.
> 
> Hi @cfernald, thanks for your reply, I think my concern is still from consumer's point of view, once the policy design keeps integrating to more and more features, and if today we have 30 features that is controlled by policy of each feature, then platform code needs to implement Policy Creators for each of the feature, the effort is a little bit big to developers that want to consume MU features, and as a consumer, we may just want to consume the feature with its default behavior, in that case the original PCDs design is even easier for consumer to consume the feature, that can reduce a lot of consumer's integrating effort when consume MU features. If we can evaluate a framework for default policies in the policy service, I think that will be the best solution, and if both Policy and PCD control design will be co-existed in MU, maybe we should also identify a best practice to suggest when to use PCD and when to use Policy design.

Agreed. For now I'm hesitant to introduce one-off default providers unless they will actually be needed to avoid establishing that pattern and instead invest in the infrastructure. I have updated the documentation, and if you have further questions or concerns about the use of policy let's discuss offline.

**Pull request merged**